internal/bguploader_test.go

```diff
@@ -185,3 +230,40 @@ func cleanup(t *testing.T, dir string) {
 		t.Log("temporary data directory was not deleted ", err)
 	}
 }
+
+// stallableDataFolder implements a portion of internal.DataFolder and can be
+// used to simulate a very slow ioutil.ReadDir operation. This can be used to
+// test BgUploader behavior in adverse conditions.
+type stallableDataFolder struct {
+	internal.DataFolder
+	dirname      string
+	ctx          context.Context
+	waitDuration time.Duration
+}
+
+func (s *stallableDataFolder) ListFilenames() ([]string, error) {
+	select {
+	case <-s.ctx.Done():
+		return nil, fmt.Errorf(
+			"fake dir reader was cancelled before wait completed: %v",
```

Comment:
TODO: for this functionality to be correct, it shouldn't return an error (that triggers a separate code path in BgUploader - either return empty list with no error or actual ReadDir results (like happy path).